tests/zquantum/core/trackers_test.py

```diff
@@ -0,0 +1,146 @@
+import json
+from functools import partial
+from os import remove
+
+import networkx as nx
+import numpy as np
+import pytest
+from qeqiskit.simulator import QiskitSimulator
+from zquantum.core.cost_function import AnsatzBasedCostFunction
+from zquantum.core.estimation import (
+    allocate_shots_uniformly,
+    estimate_expectation_values_by_averaging,
+)
+from zquantum.core.trackers import MeasurementTrackingBackend
+from zquantum.optimizers.scipy_optimizer import ScipyOptimizer
+from zquantum.qaoa.ansatzes.farhi_ansatz import QAOAFarhiAnsatz
+from zquantum.qaoa.problems.maxcut import MaxCut
+
+
+@pytest.fixture
+def H():
+    G = nx.Graph()
+    G.add_nodes_from([0, 1, 2, 3])
+    G.add_edge(0, 1, weight=10)
+    G.add_edge(0, 3, weight=10)
+    G.add_edge(1, 2, weight=1)
+    G.add_edge(2, 3, weight=1)
+    problem = MaxCut()
+    return problem.get_hamiltonian(G)
+
+
+class TestMeasurementTrackingBackend:
+    def test_save_measurement_data_from_maxcut_qaoa(self, H):
+        # Given
+        ansatz = QAOAFarhiAnsatz(1, cost_hamiltonian=H)
+        backend = MeasurementTrackingBackend(
+            inner_backend=QiskitSimulator(device_name="aer_simulator")
+        )
+
+        estimation_method = estimate_expectation_values_by_averaging
+        shot_allocation = partial(allocate_shots_uniformly, number_of_shots=100)
+        estimation_preprocessors = [shot_allocation]
+        optimizer = ScipyOptimizer(method="L-BFGS-B", options={"maxiter": 10})
+        cost_function = AnsatzBasedCostFunction(
+            H, ansatz, backend, estimation_method, estimation_preprocessors
+        )
+        initial_params = np.array([0, 0])
+        # When
+        opt_results = optimizer.minimize(cost_function, initial_params)
+        circuit = ansatz.get_executable_circuit(opt_results.opt_params)
+        backend.n_samples = 10000
+        backend.run_circuit_and_measure(circuit, n_samples=1000)
```

Comment:
If we can pass `n_samples` to `run_circuit_and_measure`, what's the reason to mutate the backend on line above it?